src/Aspire.Dashboard/Model/DashboardClient.cs

```diff
@@ -0,0 +1,309 @@
+
+                return new Uri(uri);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException($"Error parsing URIs from environment variable '{variableName}'.", ex);
+            }
+        }
+    }
+
+    public string ApplicationName => _applicationName ?? "";
+
+    ResourceViewModelSubscription IDashboardClient.SubscribeResources()
+    {
+        EnsureInitialized();
+
+        lock (_lock)
+        {
+            // There are two types of channel in this class. This is not a gRPC channel.
+            // It's a producer-consumer queue channel, used to push updates to subscribers
+            // without blocking the producer here.
+            var channel = Channel.CreateUnbounded<ResourceViewModelChange>(
+                new UnboundedChannelOptions { AllowSynchronousContinuations = false, SingleReader = true, SingleWriter = true });
+
+            ImmutableInterlocked.Update(ref _outgoingChannels, static (set, channel) => set.Add(channel), channel);
+
+            return new ResourceViewModelSubscription(
+                InitialState: _resourceByName.Values.ToImmutableArray(),
+                Subscription: StreamUpdates());
+
+            async IAsyncEnumerable<ResourceViewModelChange> StreamUpdates()
+            {
+                try
+                {
+                    await foreach (var batch in channel.Reader.ReadAllAsync(_cts.Token).ConfigureAwait(false))
+                    {
+                        yield return batch;
+                    }
+                }
+                finally
+                {
+                    ImmutableInterlocked.Update(ref _outgoingChannels, static (set, channel) => set.Remove(channel), channel);
+                }
+            }
+        }
+    }
+
+    public async IAsyncEnumerable<IReadOnlyList<(string Content, bool IsErrorMessage)>>? SubscribeConsoleLogs(string resourceName, [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        EnsureInitialized();
+
+        var client = await _client.Task.ConfigureAwait(false);
+
+        var call = client.WatchResourceConsoleLogs(
+            new WatchResourceConsoleLogsRequest() { ResourceName = resourceName },
+            cancellationToken: cancellationToken);
+
+        await foreach (var response in call.ResponseStream.ReadAllAsync(cancellationToken: cancellationToken))
+        {
+            var logLines = new (string Content, bool IsErrorMessage)[response.LogLines.Count];
+
+            for (var i = 0; i < logLines.Length; i++)
+            {
+                logLines[i] = (response.LogLines[i].Text, response.LogLines[i].IsStdErr);
+            }
+
+            yield return logLines;
+        }
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        if (Interlocked.Exchange(ref _state, StateDisposed) is not StateDisposed)
+        {
+            await _cts.CancelAsync().ConfigureAwait(false);
```

Comment:
Dispose should dispose the channel. That ensures current gRPC calls are stopped and TCP connections are closed.